File: haskell/src/Chapter1/IsUnique.hs
-- | Example of a library file. It is also used for testing the test suites.
module Chapter1.IsUnique
  (
    -- * Exported functions
    isUnique
  ) where

import qualified Data.HashMap.Strict as HM

-- | Check if String has all unique characters
--
--  >>> isUnique "uniqe ya"
--  True
--  >>> isUnique "unique ya"
--  False
--
--  Properties:
--
--  Not sure why this complains: prop> isUnique xs ==> (nub xs == xs)
--
isUnique :: String -> Bool
isUnique = fst . foldr fn (True, HM.empty)
  where
    fn c (uniq, hm) =
      if uniq && not (HM.member c hm)
      then (True, HM.insert c () hm)
      else (False, HM.empty)

File: rust/src/chapter1/is_unique.rs
// Implement an algorithm to determine if a string has all unique characters.
// What if you cannot use additional data structures?

// Thought Process:
// A: sort first, then check adjacent -> O(n log n + n) == O(n log n)
// B: iterate across string, checking or appending to "seen" list -> O(n)

use std::collections::HashMap;

fn is_unique(s: &str) -> bool {
    // Store seen characters in hash table
    //   O(s) outer loop where s is length of string
    //     O(1) hash inserts and lookups (amortized)
    //
    // O(s) total
    let mut seen = HashMap::new();
    s.chars().fold(true, |uniq, c| {
        if !uniq || seen.contains_key(&c) {
            return false;
        }
        seen.insert(c, true);
        true
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_unique() {
        assert_eq!(is_unique("i am not unique"), false);
        assert_eq!(is_unique("me tho"), true);
    }
}
